File: base/android/java/src/org/chromium/base/memory/JavaHeapDumpGenerator.java
// Copyright 2019 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.base.memory;

import android.os.Debug;

import org.jni_zero.CalledByNative;

import org.chromium.base.Log;

import java.io.IOException;

/** Enables the generation of hprof files from heap dumps. */
public final class JavaHeapDumpGenerator {
    private static final String TAG = "JavaHprofGenerator";

    private JavaHeapDumpGenerator() {}

    /**
     * Generates an hprof file at the given filePath.
     *
     * @return whether or not the hprof file was properly generated.
     */
    @CalledByNative
    public static boolean generateHprof(String filePath) {
        try {
            Debug.dumpHprofData(filePath);
        } catch (IOException e) {
            Log.i(TAG, "Error writing to file " + filePath + ". Error: " + e.getMessage());
            return false;
        }
        return true;
    }
}

File: base/android/java_heap_dump_generator.cc
// Copyright 2019 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "base/android/java_heap_dump_generator.h"

#include <jni.h>

#include <string_view>

#include "base/android/jni_string.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "base/memory_jni/JavaHeapDumpGenerator_jni.h"

namespace base {
namespace android {

bool WriteJavaHeapDumpToPath(std::string_view filePath) {
  JNIEnv* env = jni_zero::AttachCurrentThread();
  return Java_JavaHeapDumpGenerator_generateHprof(
      env, base::android::ConvertUTF8ToJavaString(env, filePath));
}

}  // namespace android
}  // namespace base
